File: Tree/definition.ts
// Definition for binary tree node
class TreeNode {
	val: number;
	left: TreeNode | null;
	right: TreeNode | null;

	constructor(
		val: number = 0,
		left: TreeNode | null = null,
		right: TreeNode | null = null
	) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

function arrayToTree(arr: (number | null)[]): TreeNode | null {
	if (!arr.length) return null;

	const root = new TreeNode(arr[0]!); // We know first element exists
	const queue: TreeNode[] = [root];
	let i = 1;

	while (queue.length && i < arr.length) {
		const node = queue.shift()!; // We know queue has elements

		// Handle left child
		if (i < arr.length && arr[i] !== null) {
			node.left = new TreeNode(arr[i]!);
			queue.push(node.left);
		}
		i++;

		// Handle right child
		if (i < arr.length && arr[i] !== null) {
			node.right = new TreeNode(arr[i]!);
			queue.push(node.right);
		}
		i++;
	}

	return root;
}

function treeToArray(root: TreeNode | null): (number | null)[] {
	if (!root) return [];

	const queue: (TreeNode | null)[] = [root];
	const level: (number | null)[] = [];

	while (queue.length) {
		const node = queue.shift()!;

		if (node !== null) {
			level.push(node.val);
			queue.push(node.left);
			queue.push(node.right);
		} else {
			level.push(null);
		}
	}

	// Remove trailing nulls
	while (level.length && level[level.length - 1] === null) {
		level.pop();
	}

	return level;
}

export { TreeNode, arrayToTree, treeToArray };

File: Tree/definition.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def arrayToTree(arr):
    if not arr:
        return None
    
    root = TreeNode(arr[0])
    queue = [root]
    i = 1
    
    while queue and i < len(arr):
        node = queue.pop(0)
        
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    
    return root

def treeToArray(root):
    # use a queue to turn the tree into its level order representation
    if not root:
        return [] 

    queue = []
    level = []
    queue = [root]
    
    while queue:
        node = queue.pop(0)

        if node is not None:
            level.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            level.append(None)

    # Remove trailing None values
    while level and level[-1] is None:
        level.pop()

    return level

